daffodil-runtime1/src/main/scala/org/apache/daffodil/processors/parsers/HexBinaryLengthParsers.scala

```diff
@@ -34,11 +34,12 @@ sealed abstract class HexBinaryLengthParser(override val context: ElementRuntime
     val dis = start.dataInputStream
     val currentElement = start.simpleElement
     val nBits = getLengthInBits(start)
-    if (nBits == 0) {
+    val nBytes = (nBits + 7) / 8
+    if (nBytes == 0) {
       currentElement.setDataValue(zeroLengthArray)
-    } else if (nBits > Int.MaxValue) {
+    } else if (nBytes > Int.MaxValue) {
```

Comment:
Although MaxValue is a technical limitation imposed by java array sizes, the way our BucketingInputSource works we still have a limitation of 256MB because that's the maximum it will store in it's cache.

If we really want to support xs:hexBinary more than 256MB, I think we could maybe use similar logic as the BlobLengthParser, which reads data in chunks. The only difference is instead of writing those chunks to a file we would copy them to an array in memory, which is eventually written to the infoset.

Thoughts? Should we include this change, or should we instead suggest that blobs be used? Do we also want a tunable limit to limit the size of blobs? And if so, do we want to default that to something less than MaxInt?

Reply:
People should have a choice of whether to use the blob files or not.
 
For a while I've been wondering if the representation of a large opaque thing should not be one hexBinary element, but an array of them. 

That plus our streaming would effectively make the size unlimited. 

Maybe users can do this themselves anyway? 

Instead of a single hexBinary element, an array of pairs of length + hexBinary of that length. Each "chunk" having a max length of say, 1024. That would multiply the effective size by 1024 which would make for well-behaved XML text, and would stream well. 
```
<xs:element name="bigData" dfdl:lengthKind="explicit" dfdl:length='{ .... }'>
   <xs:complexType>
      <xs:sequence>
         <xs:element name="data" maxOccurs="unbounded" dfdl:occursCountKind="implicit">
             <xs:complexType>
               <xs:choice>
                   <xs:element name="full"  type="xs:hexBinary" dfdl:lengthKind="explicit" dfdl:length="1024"/>
                   <!-- wants to be lengthKind 'endOfParent' below, but we don't have that in Daffodil yet -->
                   <xs:element name="final" type="xs:hexBinary" dfdl:lengthKind="pattern"
                               dfdl:lengthPattern='.{1,1023}' />
               </xs:choice>
            </xs:complexType>
         </xs:element>
       </xs:sequence>
    </xs:complexType>
  </xs:element>
```
I'm a bit unsure about the full vs. final choice and using lengthKind pattern. We've done something like this before where things extended to the end of a enclosing length-bounded structure. I just can't recall the trick.

Reply:
I think the "final" element should perhaps just have lengthKind 'delimited' but no terminator specified. Then it will extend to "end of data" which is this case is bounded by the containing structure. 

As usual with lengthKind pattern, I am missing a dfdl:assert that insists on the actual length being > 0. It might not be needed in this case because the "no forward progress" check would terminate the array. But it doesn't hurt.

Reply:
I've pushed an update that reads large xs:hexBinary elements so it at least gets around the cache issue throwing away data before it can be read into the array. So the limit should now be 2GB. Either of your solutions are probably what's required if we want to support hexBinary larger than that. I'll leave xs:hexBinary being represented as an array of arrays to another PR if we decide that's needed. That's probably a decent amount of work to update all the locations where we use hex binary correctly, and the schema solution would probably be decent workaround until then.